src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,387 @@
+                (X509EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                so = kp.getPublic();
+            }
+        }
+
+        /*
+         * KeySpec use getKeySpec after the Key has been generated.  Even though
+         * returning a binary encoding after the Base64 decoding is ok when the
+         * user wants PKCS8EncodedKeySpec, generating the key verifies the
+         * binary encoding and allows the KeyFactory to use the provider's
+         * KeySpec()
+         */
+
+        if ((EncodedKeySpec.class).isAssignableFrom(tClass) &&
+            so instanceof Key key) {
+            try {
+                // unchecked suppressed as we know tClass comes from KeySpec
+                // KeyType not relevant here.  We just want KeyFactory
+                if ((PKCS8EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                    getKeyFactory(key.getAlgorithm()).
+                        getKeySpec(key, PKCS8EncodedKeySpec.class);
+                } else if ((X509EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                    getKeyFactory(key.getAlgorithm())
+                        .getKeySpec(key, X509EncodedKeySpec.class);
+                } else {
+                    throw new IllegalArgumentException("Invalid KeySpec.");
+                }
+            } catch (InvalidKeySpecException e) {
+                throw new IllegalArgumentException("Invalid KeySpec " +
+                    "specified (" + tClass.getName() +") for key (" +
+                    key.getClass().getName() +")", e);
+            }
+        }
+
+        return tClass.cast(so);
+    }
+
+
+    private KeyFactory getKeyFactory(String algorithm) {
+        try {
+            if (factory == null) {
+                return KeyFactory.getInstance(algorithm);
+            }
+            return KeyFactory.getInstance(algorithm, factory);
+        } catch(GeneralSecurityException e){
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    // Convenience method to avoid provider getInstance checks clutter
+    private CertificateFactory getCertFactory(String algorithm) {
+        try {
+            if (factory == null) {
+                return CertificateFactory.getInstance(algorithm);
+            }
+            return CertificateFactory.getInstance(algorithm, factory);
+        } catch (GeneralSecurityException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Configures and returns a new {@code PEMDecoder} instance from the
+     * current instance that will use Factory classes from the specified
+     * {@link Provider}.  Any errors using the {@code provider} will occur
+     * during decoding.
+     *
+     * <p>If {@code params} is {@code null}, a new instance is returned with
+     * the default provider configuration.
+     *
+     * @param provider the Factory provider.
+     * @return a new PEM decoder instance.
```

Comment:
The return spec for this method and the next one should be using a consistent wording.